File: LeetCode/535 Encode and Decode TinyURL.js
var map = {};

/**
 * Encodes a URL to a shortened URL.
 *
 * @param {string} longUrl
 * @return {string}
 */
var encode = function(longUrl) {
    
    let length = 6;
    let shortUrlSuffix = '';
    let characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    let charactersLength = characters.length;
    
    for ( let i = 0; i < length; i++ ) {
        shortUrlSuffix += characters.charAt(Math.floor(Math.random() * charactersLength));
    }
    
    map[shortUrlSuffix] = longUrl;

    return `http://tinyurl.com/` + shortUrlSuffix; 
    
};

/**
 * Decodes a shortened URL to its original URL.
 *
 * @param {string} shortUrl
 * @return {string}
 */
var decode = function(shortUrl) {
    
    let length = 6;
    let shortUrlSuffix = shortUrl.substring(shortUrl.length - length);
    
    return map[shortUrlSuffix];
    
};

/**
 * Your functions will be called as such:
 * decode(encode(url));
 */

File: LeetCode/535 Encode and Decode TinyURL.java
public class Codec {
    
    static HashMap<String, String> map = new HashMap<>();

    // Encodes a URL to a shortened URL.
    public String encode(String longUrl) {
        
        String alphaNumericString = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
                                    + "0123456789"
                                    + "abcdefghijklmnopqrstuvxyz"; 
  
        StringBuilder sb = new StringBuilder(6); 
  
        for (int i = 0; i < 6; i++) { 
            int index = (int)(alphaNumericString.length() * Math.random()); 
            sb.append(alphaNumericString.charAt(index)); 
        } 
  
        String shortUrlSuffix = sb.toString();
        
        map.put(shortUrlSuffix, longUrl);
        
        return "http://tinyurl.com/" + shortUrlSuffix;
        
    }

    // Decodes a shortened URL to its original URL.
    public String decode(String shortUrl) {
        
        String shortUrlSuffix = shortUrl.substring(shortUrl.length() - 6);
        
        return map.get(shortUrlSuffix);
        
    }
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.decode(codec.encode(url));